usr/sbin/pkcsslotd/mutex.c

```diff
@@ -27,11 +27,16 @@ int CreateXProcLock(void)
     struct group *grp;
     mode_t mode = (S_IRUSR | S_IRGRP);
 
+    /*
+    * Linux implementation of flock doesn't care about the mode the file
+    * is opened in, but AIX does - to obtain exclusive locks, AIX needs
+    * the file to be opened in write mode. Minority wins.
+    */
     if (xplfd == -1)
-        xplfd = open(OCK_API_LOCK_FILE, O_RDONLY);
+        xplfd = open(OCK_API_LOCK_FILE, O_WRONLY);
```

Comment:
The reason for the EPERM is certainly this:
`mode_t mode = (S_IRUSR | S_IRGRP);` so the lock file has `-r--r-----`. Certainly, it is not allowed to open this file for write ....

We could change the mode to match `O_WRONLY`, but then we still might have left over lock files from previous installations with `-r--r-----` which then could not be opened properly when using `O_WRONLY`. This could create migration problems when upgrading from older versions. 

So again, leave it as is (`O_RDONLY`) for Linux, but you may change it for AIX. You might need to adjust the file mode as well for AIX....

And this applies the same for all lock files used by opencryptoki.

Reply:
We _could_ do a best-effort chmod before the file `open` operation is attempted. That would prevent yet another platform-specific `#ifdef` - I've currently factored the file mode and the open mode into their distinct preprocessor `#define`, setting values according to the platform.

Trying a `chmod` automatically handles migration as well.

Reply:
a `chmod` may be ok for pkcsslord's lock file, but possibly not for the other lock files. The other lock files might have been created by other users, so the current user might not be able to do a chmod on them.

Reply:
well, even for pkcsslotd's lockfile: There was a time (not sooo long ago) where pkcsslotd was running as root, and thus the lock file owner was root, too. If one still has such a lock file owned by root laying around, then a chmod done by the pkcsslotd user won't work either.

So I guess, even though it adds even more #ifdefs, the best solution probably is to keep it O_RDONLY for Linux, and make it O_WRONLY for AIX only.

Reply:
I've changed the modes to `u+w,g+w` on AIX (and left Linux unchanged). Linux now opens the file as `O_RDONLY` and AIX opens the file as `O_WRONLY`.

Manually killing the daemon (`kill -KILL ...`) and trying to restart it with the API lock file still around is not a problem anymore.